### PR TITLE
PP-325/REL-002

### PR DESCRIPTION
**Pull request by franciscotobar**
## What

- The problem is related to the fact that the initialize method of the template contract is never called, so a malicious attacker could call it and therefore set a logic address. After doing that, the attacker could call any function and that would lead the logic address to be executed. Potentially the logic address could contain a selfdestruct instruction that would destroy the template contract. If so, the proxies contracts won’t be able to execute any transaction anymore.

- I decided to just initialise the CustomSmartWallet implementation contract in the constructor.

## Why

- By using this approach the CustomSmartWallet initialization is blocked since the deployment to prevent malicious initialisation.

## Refs

- [Jira issue](https://rsklabs.atlassian.net/browse/PP-325)

### CODE REVIEW

File: contracts/smartwallet/CustomSmartWallet.sol [hunk 1]
```diff
@@ -14,7 +14,11 @@ contract CustomSmartWallet is IForwarder {
 
     uint256 public override nonce;
     bytes32 public constant DATA_VERSION_HASH = keccak256("2");
-    bytes32 public domainSeparator;    
+    bytes32 public domainSeparator;
+
+    constructor() public {
+        setOwner(msg.sender);
```
Comment (ironFe93) [inline contracts/smartwallet/CustomSmartWallet.sol:20]: Wouldn't caling setOwner() here make it impossible to run initialize() afretwards?

**Review by ironFe93 — COMMENTED** [review]
left a comment, if responded I can approve.

File: contracts/smartwallet/CustomSmartWallet.sol (hunk 1 above)
Comment (franciscotobar) [reply contracts/smartwallet/CustomSmartWallet.sol:20]: By initializing it, we are initializing the logic contract and not the proxy which in our case are the smart wallet address created with the index. 

Im still doing to testing to make sure that its working properly but so far so good. I just need to confirm if we need to setup to msg.sender or a different address or if we add an initialization flag, but that increase the gas cost.

**Comment by franciscotobar** [comment]
I tried to use expect and assert, specifically the expect().to.be.rejectedWith and assert.isRejected. It seems to be that it cannot be combined.

**Review by antomor — APPROVED** [review]
@franciscotobar Thank you for these changes, LGTM!

**Pull request merged**